### PR TITLE
fix(cognito): tighten domainPrefix validation to match AWS constraint

### LINKED ISSUE #38531: (aws-cognito): domainPrefix validation is looser than the documented Domain constraint
### Describe the bug

`UserPoolDomain` validates `cognitoDomain.domainPrefix` against `/^[a-z0-9-]+$/`.

The documented constraint for `AWS::Cognito::UserPoolDomain.Domain` is narrower: pattern `^[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?$`, minimum length 1, maximum length 63.

CDK's pattern accepts everything the documented one does and more, so a few shapes pass synth and get written into the template unchanged:

- a leading hyphen, `-myprefix`
- a trailing hyphen, `myprefix-`
- hyphens only, `---`
- anything longer than 63 characters

The check in `packages/aws-cdk-lib/aws-cognito/lib/user-pool-domain.ts` is the only synth-time validation the prefix gets, and the value flows straight through to the L1 `Domain` property.

### Expected Behavior

Synth fails for a prefix the documented pattern rejects, the same way it already does for an uppercase or underscore-bearing one.

### Current Behavior

Synth succeeds and the value lands in the template as written:

```
leading hyphen      | SYNTH OK | Domain="-myprefix"
trailing hyphen     | SYNTH OK | Domain="myprefix-"
all hyphens         | SYNTH OK | Domain="---"
length 64           | SYNTH OK | Domain="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
control (valid)     | SYNTH OK | Domain="valid-prefix-123"
control (uppercase) | THREW    | domainPrefix for cognitoDomain can contain only lowercase alphabets, numbers and hyphens
```

The uppercase control is there to confirm validation actually runs, so these four are getting through the check rather than skipping it.

### Reproduction Steps

```ts
import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as cognito from 'aws-cdk-lib/aws-cognito';

for (const prefix of ['-myprefix', 'myprefix-', '---', 'a'.repeat(64)]) {
  const stack = new Stack(new App(), 'S');
  const pool = new cognito.UserPool(stack, 'Pool');

  // expected: throws, since the documented pattern rejects all four
  // actual: synthesizes, and the value lands in the template unchanged
  pool.addDomain('Domain', { cognitoDomain: { domainPrefix: prefix } });

  console.log(prefix, JSON.stringify(Template.fromStack(stack).toJSON().Resources));
}
```

### Possible Solution

Match the documented constraint:

```ts
!/^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/.test(props.cognitoDomain.domainPrefix)
```

That covers the length bound too, so it needs no separate check, and it keeps the existing error code. The message would want rewording, since "can contain only lowercase alphabets, numbers and hyphens" no longer describes what is being enforced.

On backwards compatibility, AGENTS.md draws the line at whether the input ever deployed. These values do not appear deployable, so by that rule this is the no-feature-flag case and failing fast at synth is preferred. That is worth a maintainer's read rather than my assumption, given the caveat below.

Happy to open a PR if this is something you want fixed.

### Additional Information/Context

Two things worth flagging so nobody takes more from this than it supports.

The pattern and the length bound live in the CloudFormation documentation. The resource schema published at `schema.cloudformation.us-east-1.amazonaws.com/aws-cognito-userpooldomain.json` types `Domain` as a plain string with no pattern and no bounds, so the rejection is presumably Cognito rather than CloudFormation validating client-side. I have not deployed any of these, so treat "these are invalid" as coming from the documentation rather than from a failed deploy I watched.

This is also not an explanation for #37514. That one is about tokens, and every value here is static.

### AWS CDK Library version (aws-cdk-lib)

2.264.0

### AWS CDK CLI version

2.1131.0

### Node.js Version

v24.12.0

### OS

Windows 11 Pro 10.0.26200

### Language

TypeScript

### PR DESCRIPTION
**Pull request by sanyamk23**
### Issue # (if applicable)

Closes #38531.

### Reason for this change

The AWS Cognito API only accepts a domain prefix that starts and ends with a lowercase alphanumeric character, contains only lowercase alphanumeric characters and hyphens (not leading or trailing), and is 1-63 characters in length. The previous validation (`/^[a-z0-9-]+$/`) accepted values with leading/trailing hyphens or longer than 63 characters, which fail at deploy time with an opaque API error.

### Description of changes

Tightened the `domainPrefix` validation in `UserPoolDomain` to match the documented CloudFormation pattern `^[a-z0-9](?:[a-z0-9\-]{0,61}[a-z0-9])?$`, so invalid prefixes are rejected at synth time with a clear error message.

Updated the existing unit tests for the new message and added cases for leading/trailing hyphens, >63 characters, and valid boundary values (1 char, hyphens in middle, exactly 63 chars). Added an integration test with snapshot exercising a valid prefix.

### Describe any new or updated prerequisites

None.

### Any behavior changes

Yes — previously accepted invalid `domainPrefix` values (leading/trailing hyphens, >63 chars) now fail at synthesis time instead of at deploy time. All existing integration test prefixes remain valid under the new pattern.

### Is this a breaking change to the CDK CLI?

No

### Is this a breaking change to the library or any generated client interfaces?

No

### CODE REVIEW

**Comment by sanyamk23** [comment]
The `cli-changes` check failed with `Unable to locate the commit sha: f89474a338fe212886306e307d188190463c5824` — a shallow-fetch depth issue in the workflow, unrelated to the code change. Could someone with write access re-run the failed job?

File: packages/@aws-cdk-testing/framework-integ/test/aws-cognito/test/integ.user-pool-domain-prefix-validation.ts
Comment (leonmk-aws) [inline packages/@aws-cdk-testing/framework-integ/test/aws-cognito/test/integ.user-pool-domain-prefix-validation.ts:1]: This tightened validation is a synth-time validation, this test is essentially the same as https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk-testing/framework-integ/test/aws-cognito/test/integ.user-pool-domain-cfdist.ts lets drop it.

FYI you can add the `pr-linter/exemption-requested` label to request a test exemption if you believe that an integration test, unit test or readme change is not required for a PR. A maintainer will then review and add exemptions label if needed. For this PR, I have added the exemption label

File: packages/aws-cdk-lib/aws-cognito/lib/user-pool-domain.ts [hunk 1]
```diff
@@ -162,8 +162,8 @@ export class UserPoolDomain extends Resource implements IUserPoolDomain {
 
     if (props.cognitoDomain?.domainPrefix &&
       !Token.isUnresolved(props.cognitoDomain?.domainPrefix) &&
-      !/^[a-z0-9-]+$/.test(props.cognitoDomain.domainPrefix)) {
-      throw new ValidationError(lit`DomainPrefixCognitoDomainContain`, 'domainPrefix for cognitoDomain can contain only lowercase alphabets, numbers and hyphens', this);
+      !/^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/.test(props.cognitoDomain.domainPrefix)) {
+      throw new ValidationError(lit`DomainPrefixCognitoDomainContain`, 'domainPrefix for cognitoDomain must start and end with a lowercase alphanumeric character, contain only lowercase alphanumeric characters and hyphens (not leading or trailing), and be 1-63 characters in length', this);
```
Comment (leonmk-aws) [inline packages/aws-cdk-lib/aws-cognito/lib/user-pool-domain.ts]: Nit: can we add the value of the domain prefix at the end of the error message ?
Something like `...characters in length, got ${props.cognitoDomain.domainPrefix}`

**Review by leonmk-aws — DISMISSED** [review]
Thank you for your contribution, 2 small comments

**Comment by Zelys-DFKH** [comment]
Context that may be useful for reviewers, not a review verdict.

I looked into whether the validation engine already covers these shapes, since that would make this change redundant. It does not, at least not today. `packages/aws-cdk-lib/package.json` pins `@aws/cloudformation-validate` at exactly `1.7.0-beta`, with no caret or tilde, and lists it under `bundleDependencies`. I diffed the shipped engine data at `src/data-source/generated/schema-validator/compiled_schemas.json` between tags: in 1.7.0-beta, `AWS::Cognito::UserPoolDomain.Domain` is a bare `{"type": "string"}` with no constraints. The length and pattern constraints first appear in 1.8.0-beta, released 2026-08-25.

So CDK does not float to newer engine releases, and the version it currently bundles has nothing to say about `Domain`. The regex in this PR is the only thing catching leading hyphens, trailing hyphens, and over-63-character prefixes for CDK users right now.

The published registry schema at `schema.cloudformation.us-east-1.amazonaws.com/aws-cognito-userpooldomain.json` also still returns a bare `{"type": "string"}` for `Domain`, so the constraint currently lives in the validate engine's data rather than in the registry.

Whether that changes anything about this PR is a maintainer call.

**Comment by sanyamk23** [comment]
Thanks for the review! I addressed both comments:

1. The error message now includes the actual domainPrefix value so users can see exactly which prefix was rejected.

2. The exemption label covers the integration test requirement.

Ready for another pass when you have a moment.
